src/charm.py

```diff
@@ -342,6 +351,71 @@ def _on_tls_certificate_pki_certificate_available(self, _: CertificateAvailableE
         """Handle the tls-certificates-pki certificate available event."""
         self._add_intermediate_ca_certificate_to_pki_secrets_engine()
 
+    def _on_vault_pki_certificate_creation_request(
+        self, event: CertificateCreationRequestEvent
+    ) -> None:
+        """Handle the vault-pki certificate creation request event."""
+        self._generate_pki_certificate_for_requirer(
+            event.certificate_signing_request, event.relation_id
+        )
+
+    def _generate_pki_certificate_for_requirer(self, csr: str, relation_id: int):
+        """Generate a PKI certificate for a TLS requirer."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-pki certificate request")
+            return
+        if not self._tls_certificates_pki_relation_created():
+            logger.debug("TLS Certificates PKI relation not created")
+            return
+        vault = self._get_vault_client()
+        if (
+            not vault
+            or not vault.is_api_available()
+            or not vault.is_initialized()
+            or vault.is_sealed()
+        ):
+            return
+        if not (approle_auth := self._get_vault_approle_secret()):
+            return
+        vault.authenticate(AppRole(approle_auth[0], approle_auth[1]))
+        common_name = self._get_config_common_name()
+        if not common_name:
+            logger.error("Common name is not set in the charm config")
+            return
+        if not vault.is_pki_role_created(role=VAULT_PKI_ROLE, mount=VAULT_PKI_MOUNT):
+            logger.debug("PKI role not created")
+            return
+        requested_csr = csr
+        requested_common_name = get_common_name_from_csr(requested_csr)
+        certificate = vault.sign_pki_certificate_signing_request(
+            mount=VAULT_PKI_MOUNT,
+            role=VAULT_PKI_ROLE,
+            csr=requested_csr,
+            common_name=requested_common_name,
+        )
+        if not certificate:
+            logger.debug("Failed to sign the certificate")
+            return
+        self.vault_pki.set_relation_certificate(
+            relation_id=relation_id,
+            certificate=certificate.certificate,
+            certificate_signing_request=csr,
+            ca=certificate.ca,
+            chain=certificate.chain,
+        )
+
+    def _sync_vault_pki(self) -> None:
+        """Goes through all the vault-pki relations and sends necessary TLS certificate."""
+        for relation in self.model.relations[PKI_RELATION_NAME]:
+            outstanding_requests = self.vault_pki.get_outstanding_certificate_requests(
+                relation_id=relation.id
+            )
+            for request in outstanding_requests:
+                self._generate_pki_certificate_for_requirer(
+                    csr=request.csr,
+                    relation_id=relation.id,
+                )
```

Comment:
I think this can be simplified to:
```suggestion
        for request in outstanding_requests:
            self._generate_pki_certificate_for_requirer(
                csr=request.csr,
                relation_id=request.relation_id,
            )
```

It should also be a bit more performant because it will do less looping.

Reply:
Agreed, done